contractcourt/chain_arbitrator.go

```diff
@@ -641,6 +641,85 @@ func (c *ChainArbitrator) Start() error {
 	return nil
 }
 
+// dispatchBlocks consumes a block epoch notification stream and dispatches
+// blocks to each of the chain arb's active channel arbitrators. This function
+// must be run in a goroutine.
+func (c *ChainArbitrator) dispatchBlocks(
+	blockEpoch *chainntnfs.BlockEpochEvent) {
+
+	// On exit, cancel our blocks subscription and shutdown all of our
+	// block queues so that the arbitrators know they will no longer be
+	// receiving blocks.
+	defer func() {
+		blockEpoch.Cancel()
+
+		c.Lock()
+		for _, channel := range c.activeChannels {
+			channel.blocks.Stop()
+		}
+		c.Unlock()
+	}()
+
+	// Consume block epochs until we receive the instruction to shutdown.
+	for {
+		select {
+		// Consume block epochs, exiting if our subscription is
+		// terminated.
+		case block, ok := <-blockEpoch.Epochs:
+			if !ok {
+				log.Trace("dispatchBlocks block epoch " +
+					"cancelled")
+
+				return
+			}
+
+			// To keep our locking simple, we track whether we broke
+			// our dispatch loop due to shutdown with this bool.
+			var shutdown bool
+
+			// Acquire our lock and dispatch this new block to each
+			// channel arbitrator, breaking early if we receive the
+			// instruction to shutdown.
+			c.Lock()
+		dispatch:
+			for _, arb := range c.activeChannels {
```

Comment:
Alternatively, you can skip the locking by just grabbing all the in chans outside the loop and sending to those directly here. Can be a map from chan point to in chan.

Reply:
Otherwise this may block a lot of other expected start up operations.